File: convolve.hs
module Main where

import System.Environment (getArgs)
import Data.List
import qualified Data.ByteString as BS

rounddiv p q = if (r*2) >= q then d+1 else d
               where d = p `div` q
                     r = p `mod` q

kernId           = [[0,0,0],[0,1,0],[0,0,0]]
kernBlur         = [[1,1,1],[1,1,1],[1,1,1]]
kernGaussianBlur = [[1,2,1],[2,4,2],[1,2,1]]
kernEdgeDetect   = [[-1,-1,-1],[-1,8,-1],[-1,-1,-1]]
kernSharpen      = [[0,-1,0],[-1,5,-1],[0,-1,0]]
kernEmboss       = [[-2,-1,0],[-1,1,1],[0,1,2]]

ksum = sum.map sum

-- image convolution with a given kernel
-- x is the neighbourhood of a pixel
-- x has the same size and shape as the kernel
--
convolve kernel w x = ((max 0).(min 255)) val
                      where val = ((`rounddiv` w) . ksum . kmul kernel) x
                            kmul = zipWith (zipWith (*))

-- convert a list into a list of 'neighbourhoods'
-- i.e. the elements to either side of each element in the original list
--
-- e.g. ns "Hello" -> ["Hel","ell","llo"]
-- e.g. ns [1,2,3,4,5,6] -> [[1,2,3],[2,3,4],[3,4,5],[4,5,6]]
--
ns [] = []
ns (x:[]) = []
ns (x:y:[]) = []
ns z = take 3 z : ns (tail z)

-- split a list into a 2-D list of lists
-- e.g. split 2 "Hello!" -> ["He","ll","o!"]
-- e.g. split 3 [1,2,3,4,5,6] -> [[1,2,3],[4,5,6]]
--
split n [] = []
split n x = take n x : (split n . drop n) x

-- create a border around the edge of our image
--
addBorder xss = (extend . map extend) xss
                where extend = extendr.extendl
                      extendr = reverse.extendl.reverse
                      extendl (x:xs) = x:x:xs

convolveImage :: [[Integer]] -> [[Integer]] -> [[Integer]]
convolveImage img ker = (map.map) (convolve ker weight) ws
                        where ys = ns img
                              zs = (map.map) ns ys
                              ws = map transpose zs
                              w = ksum ker
                              weight = if w==0 then 1 else w

-- 32 bits little endian at offset 0x12
getBitmapHeaderWidth bs = b0+b1*256+b2*256*256+b3*256*256*256
                          where b0 = fromIntegral(BS.index bs 18)
                                b1 = fromIntegral(BS.index bs 19)
                                b2 = fromIntegral(BS.index bs 20)
                                b3 = fromIntegral(BS.index bs 21)

getKernel "blur"    = kernBlur
getKernel "gblur"   = kernGaussianBlur
getKernel "edge"    = kernEdgeDetect
getKernel "sharpen" = kernSharpen
getKernel "emboss"  = kernEmboss
getKernel _         = kernId

main :: IO ()
main = do
    args <- getArgs
    let inFile = head args
    contents <- BS.readFile inFile
    let kern = last args
    let header = BS.take 1078 contents
    let width = getBitmapHeaderWidth header
    let pixelData = BS.drop 1078 contents
    let imagePixels = (addBorder . split width . map toInteger . BS.unpack) pixelData
    let newImage = convolveImage imagePixels (getKernel kern)
    let bytes = BS.append header (BS.pack ((map fromInteger . concat) newImage))
    let outFile = takeWhile (/='.') inFile ++ '_':kern ++ (dropWhile (/='.') inFile)
    BS.writeFile outFile bytes

File: convolve.py
import sys
import argparse
from funclist import head, tail, take, drop, transpose, zipWith, mapmap, concat

kernId           = [[0,0,0],[0,1,0],[0,0,0]]
kernBlur         = [[1,1,1],[1,1,1],[1,1,1]]
kernGaussianBlur = [[1,2,1],[2,4,2],[1,2,1]]
kernEdgeDetect   = [[-1,-1,-1],[-1,8,-1],[-1,-1,-1]]
kernSharpen      = [[0,-1,0],[-1,5,-1],[0,-1,0]]
kernEmboss       = [[-2,-1,0],[-1,1,1],[0,1,2]]

# image convolution with a given kernel
# x is the neighbourhood of a pixel
# x has the same size and shape as the kernel
#
mul = lambda x: lambda y: x*y

def ksum(xss):
    return sum(list(map(sum, xss)))

def convolve1(kernel, weight, x):
    kmul = zipWith(zipWith(mul))
    val = ksum(kmul(kernel)(x)) // weight
    return max(0, min(255, val))

convolve = lambda k: lambda w: lambda x: convolve1(k, w, x)

# convert a list into a list of 'neighbourhoods'
# i.e. the elements to either side of each element in the original list
#
# e.g. ns "Hello" -> ["Hel","ell","llo"]
# e.g. ns [1,2,3,4,5,6] -> [[1,2,3],[2,3,4],[3,4,5],[4,5,6]]
#
def ns(xs):
    if len(xs)<3:
        return []
    else:
        return [take(3)(xs)]+ns(tail(xs))

# split a list into a 2-D list of lists
# e.g. split 2 "Hello!" -> ["He","ll","o!"]
# e.g. split 3 [1,2,3,4,5,6] -> [[1,2,3],[4,5,6]]
#
def split1 (n, xs):
    if len(xs) == 0:
        return []
    else:
        return [take(n)(xs)]+split1(n, drop(n)(xs))

split = lambda n: lambda xs: split1(n, xs)

# create a border of 0's around the edge our image
#
def addBorder(xss):
    yss = list(map(lambda xs: [0]+xs+[0], xss))
    l = len(yss[0])
    return [l*[0]] + yss + [l*[0]]

def convolveImage(img, ker):
    w = ksum(ker)
    if w == 0:
        w = 1
    ys = ns(img)
    zs = mapmap(ns)(ys)
    ws = list(map(transpose, zs))
    return mapmap(convolve(ker)(w))(ws)

# 32 bits little endian at offset 0x12
def getBitmapHeaderWidth(bs):
    b0 = bs[18]
    b1 = bs[19]
    b2 = bs[20]
    b3 = bs[21]
    return b0+b1*256+b2*256*256+b3*256*256*256

def getKernel(kern):
    if kern == "blur":
        return kernBlur
    if kern == "gblur":
        return kernGaussianBlur
    if kern == "edge":
        return kernEdgeDetect
    if kern == "sharpen":
        return kernSharpen
    if kern == "emboss":
        return kernEmboss
    return kernId

def parse_command_line():
    parser = argparse.ArgumentParser(description="Convolution")
    parser.add_argument("-f", metavar="<filename>", help="path/filename of image")
    parser.add_argument("-k", metavar="<kernel>", help="kernel: edge, blur, emboss, etc.")
    args = parser.parse_args()
    return args

def main():
    sys.setrecursionlimit(3000)
    args = parse_command_line()
    inFile = args.f
    try:
        with open(inFile, 'rb') as file:
            contents = file.read()

    except (FileNotFoundError, PermissionError) as err:
        sys.stderr.write(str(err) + "\n")
        sys.stderr.write("Could not open " + inFile + "\n")

    except IOError as err:
        sys.stderr.write(str(err) + "\n")
        sys.stderr.write("Error reading " + inFile + "\n")

    kern = args.k
    header = take(1078)(contents)
    width = getBitmapHeaderWidth(header)
    pixelData = drop(1078)(contents)
    imagePixels = addBorder(split(width)(list(pixelData)))
    newImage = convolveImage(imagePixels, getKernel(kern))
    imagebytes = header + bytearray(concat(newImage))
    #let outFile = takeWhile (/='.') inFile ++ '_':kern ++ (dropWhile (/='.') inFile)
    outFile = open('outfile.bmp', 'wb')
    outFile.write(imagebytes)
    outFile.close()

if __name__ == "__main__":
    main()